### PR TITLE
ci: add nightly OpenAI API surface tests (P1-1205)

### LINKED ISSUE #5: [Feature] Support page size greater than one
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

### Motivation

- Flash attention performs poorly when page size = 1. To improve inference performance, it is necessary to support page size > 1.

### Related resources

_No response_

### LINKED ISSUE #1205: [Nightly] Section 4.A: OpenAI-compatible API surface coverage
> Sub-issue of #1117. Implements section 4.A.

### Objective
Daily nightly coverage for OpenAI-compatible API surface beyond basic completion,
modeled on vllm `tests/entrypoints/openai/` + sglang `ToolCallTestParams`.

### Gating
**Mandatory** — failure blocks nightly.
Slash command: `/run-nightly api.openai`.

### Tasks (P0)
- [ ] Reusable `OpenAIApiTestParams` runner abstraction
- [ ] `test_basic`: `tools=[add(a,b)]`, validate `tool_calls` field
- [ ] `test_auto`: `tool_choice="auto"`
- [ ] `test_streaming`: SSE chunk reassembly
- [ ] `test_multiturn`: append `tool_response` and continue
- [ ] Wire into nightly 1-tpu workflow

### Tasks (P1, separate PRs)
- [ ] `test_required` / `test_none` / `test_specific` / `test_strict` / `test_parallel`
- [ ] `test_thinking` (reasoning-mode models)
- [ ] Request validation (4xx for malformed)
- [ ] `logprobs` / `n>1` / `best_of` / `response_format=json_object`

### Dependencies
- Small model: Qwen3-1.7B + `--tool-call-parser=qwen3_coder` (confirm availability)
- OpenAI server must support: `tools=[...]`, `tool_calls`, SSE, multi-turn

### Contributor checklist (per subtest PR)
- [ ] **Threshold rationale**: pass criteria (typically 0/1 binary, but document edge cases)
- [ ] **Estimated wall-clock**: per subtest
- [ ] **Stability**: deterministic? if not, document flake mitigation

### Acceptance criteria
- All P0 subtests run nightly, total ≤ 10 min on 1-tpu
- Per-subtest pass count in GH summary

### LINKED ISSUE #1272: [Bug] json_schema constraint not applied: dict passed instead of string in _build_sampling_params
## Description

In `serving_chat.py`, `_build_sampling_params` passes the `json_schema` constraint value as a Python dict to the grammar backend, but the backend expects a serialized JSON string. As a result, constrained decoding silently does nothing — the model generates unconstrained output even when `tool_choice="required"` or a specific function is requested.

## Root Cause

The constraint value from `FunctionCallParser.get_structure_constraint()` returns a dict for `json_schema` type, but the downstream `GuidanceBackend.dispatch_json()` expects a string argument (`key_string`). Without `convert_json_schema_to_str()`, the dict is passed through and the grammar backend ignores it.

## Impact

- `tool_choice="required"` does not actually constrain decoding — model may produce free text instead of tool calls
- `tool_choice={"type":"function","function":{"name":"..."}}` similarly unconstrained
- Only affects json_schema constraint path, not structural_tag

## Fix

Call `convert_json_schema_to_str()` on the constraint value before passing to the grammar backend.

Fixed in #1247.

### PR DESCRIPTION
**Pull request by Rodrian7**
## Summary

- Add P0 + P1 tool-calling API surface tests for `/v1/chat/completions` to nightly daily CI
- Implements `OpenAIApiTestParams` frozen reusable runner abstraction + 13 test cases covering tool calling, streaming, tool_choice variants, thinking mode, logprobs, and n>1
- Per-subtest pass/fail count written to `$GITHUB_STEP_SUMMARY` (issue acceptance criteria)
- Wires into `nightly-test-daily.yml` as a new 1-TPU job with `nightly-test-daily-finish` summary integration

### Pre-existing bug fixes

1. **json_schema constraint not applied** — `_build_sampling_params` passed the constraint value as a dict, but the grammar backend expected a string. Constrained decoding silently did nothing.
   - Fix: call `convert_json_schema_to_str()` before passing to backend.

2. **Thinking mode breaks structural_tag grammar** — when `structural_tag` constraint is active, `<think>` tokens violate the grammar and cause a parser error.
   - Fix: auto-set `enable_thinking=False` when structural_tag is applied. Note: only effective on the Jinja template path (`_apply_jinja_template`); `_apply_conversation_template` does not read `chat_template_kwargs` (pre-existing gap).

3. **Bare JSON not parsed as tool calls** — with `tool_choice="required"`, json_schema constraint forces bare JSON output (no `<tool_call>` tags), but the detector only recognizes tagged format, leaving `tool_calls=None`.
   - Fix: add `orjson.loads` → `parse_base_json` fallback in `_process_tool_calls`.

4. **Grammar error crashes scheduler** (**behavior change**) — `fill_vocab_mask` called `_check_err()` which raises `ValueError` on grammar error, propagating up and killing the server process. Additionally, `accept_token` now raises `ValueError` on grammar error (old code silently set `finished=True`), causing the scheduler to abort the request instead of silently completing it. Both changes align with sglang main.
   - Fix: `fill_vocab_mask` replaced with `logger.warning` + `self.finished = True`; `accept_token` sets `self.finished = True` before raising.

5. **`is_terminated` misaligned with sglang main** (**behavior change**) — returned `ll_matcher.is_stopped()` (grammar reaches accept state → immediate stop) instead of `self.finished` (requires EOS token or error). This affects all constrained decoding paths (json_schema, regex, ebnf, structural_tag). Changed to return `self.finished` to align with sglang main. Verify e2e constraint suite (test_json_mode, test_ebnf, test_structural_tag) is green before merge.
   - Fix: return `self.finished` instead of `ll_matcher.is_stopped()`.

6. **`grammar.finished` overwritten after error** — after `accept_token` raises, `req.grammar.finished = req.finished()` resets it from `True` back to `False`, causing the errored grammar to keep receiving tokens.
   - Fix: move the assignment into the `else` clause of the try/except.

7. **`_get_tool_schema` missing `"type": "object"`** — JSON Schema for tool calls lacked explicit type, causing llguidance to compile a permissive grammar that allows non-object array items (e.g. strings). `tool_choice=specific` was broken because its schema used `_get_tool_schema` directly as `items`.
   - Fix: add `"type": "object"` to `_get_tool_schema` return value.

Closes #1205. Fixes #1272, #1273, #1274, #1281. Refs: #1117.

## Supported cases

| Case | Tool | tool_choice | Stream | Pass criteria |
|---|---|---|---|---|
| `test_basic` | `add(a,b)` | default | no | `tool_calls` non-null, function name = "add", args parse as JSON with `a` and `b` |
| `test_auto` | `add(a,b)` | `"auto"` | no | Same as basic, verifies explicit `tool_choice="auto"` path |
| `test_streaming` | `add(a,b)` | default | yes | SSE chunks reassemble into complete tool call with valid id, name, arguments |
| `test_multiturn` | `add(a,b)` | default | no | First turn produces tool call; second turn (after tool result) produces assistant content |
| `test_required` | `add(a,b)` | `"required"` | no | Model forced to call a tool even on non-tool-related prompt |
| `test_none` | `add(a,b)` | `"none"` | no | `tool_calls` is None despite tool-directed prompt; content is non-null |
| `test_specific` | `add(a,b)` | `{"type":"function","function":{"name":"add"}}` | no | Model calls the named function; function name == "add" |
| `test_strict` | `add(a,b)` strict | default | no | `strict: true` + `additionalProperties: false` schema produces valid tool call |
| `test_parallel` | `add(a,b)` | `"required"` | no | Prompt asks for 2 additions; if `len(tool_calls) < 2`, skipTest |
| `test_thinking` | — | — | no | `enable_thinking` via `chat_template_kwargs`; skipTest on 4xx, re-raise on 5xx |
| `test_malformed_request_4xx` | malformed | — | no | Invalid tool schema returns 4xx, not 500 |
| `test_logprobs` | — | — | no | `logprobs=True, top_logprobs=3`; validate per-token probability structure |
| `test_n_greater_than_1` | — | — | no | `n=2` returns exactly 2 choices with valid content |

All tool-calling cases use `temperature=0` and directive prompts for determinism.

## Launch command

Server is launched by `popen_launch_server` in `setUpClass`:

```
python3 -m sgl_jax.launch_server \
  --model-path Qwen/Qwen3-1.7B \
  --trust-remote-code \
  --skip-server-warmup \
  --dtype bfloat16 \
  --mem-fraction-static 0.8 \
  --max-running-requests 16 \
  --page-size 64 \
  --random-seed 42 \
  --tool-call-parser qwen25 \
  --device tpu \
  --api-key sk-123456
```

## Test command

```bash
cd test/srt
python3 run_suite.py --suite nightly-test-openai-api-surface-tpu-v6e-1-daily
```

Or directly:

```bash
cd test/srt
python3 -m unittest openai_server.nightly.test_openai_api_surface
```

## Estimated wall-clock

| Phase | Time |
|---|---|
| Server launch + JAX precompile | ~3-5 min |
| P0 tests (basic, auto, streaming, multiturn) | ~2 min |
| P1 tests (9 methods) | ~3 min |
| **Total** | **~8-10 min** |

CI timeout set to 15 min to allow headroom for cold cache. Issue acceptance criteria requires total ≤ 10 min.

## Threshold rationale

All tests are binary pass/fail. No accuracy thresholds — each test validates structural correctness of the API response:
- **P0 tests**: `tool_calls` populated, function name matches, SSE reassembly, multi-turn roundtrip
- **P1 tool_choice**: `required` forces tool call, `none` suppresses it, specific dict targets function, `strict` uses JSON schema constraint
- **P1 parallel**: skipTest if model doesn't produce ≥2 calls (model capability, not server bug)
- **P1 thinking**: skipTest on 4xx (model may not support thinking mode); 5xx re-raised as test failure
- **P1 validation**: server returns 4xx (not 500) for malformed tool schema
- **P1 logprobs**: per-token probability structure with 3 top_logprobs
- **P1 n>1**: exactly 2 choices returned

## Stability

**Medium risk.** Qwen3-1.7B's tool-calling ability with `qwen25` parser has not been validated at scale in this repo's nightly CI. The existing `test_tool_calls.py` uses the larger Qwen3-Coder-30B-A3B on 4 TPUs. Small models may occasionally fail to produce well-formed tool calls. Mitigations in place:
1. `temperature=0` + directive prompts ("Use the add tool") for maximum determinism
2. `CustomTestCase` already supports `SGLANG_TEST_MAX_RETRY` env var for CI retries
3. `qwen25` parser confirmed available in `function_call_parser.py`
4. `test_parallel` and `test_thinking` use `skipTest` fallback for model capability gaps

If Qwen3-1.7B proves unstable, the model can be upgraded without changing test structure (just `DEFAULT_SMALL_MODEL_NAME_FOR_TEST`).

## What's intentionally not in this PR

- `best_of` parameter testing
- `response_format=json_object` structured output
- `parallel_tool_calls` protocol field (not implemented server-side)
- Streaming + `tool_choice="required"`/specific combination — bare-JSON fallback only implemented for non-streaming path (see #1275)
- Streaming + n>1 combination
- Multi-model matrix (only Qwen3-1.7B)

## Files changed

**Production code:**
- `python/sgl_jax/srt/constrained/llguidance_backend.py` — fix `GuidanceGrammar` lifecycle: `accept_token` EOS handling, graceful error handling in `fill_vocab_mask`, `is_terminated` returns `self.finished` (**behavior change** — see bug fix #5)
- `python/sgl_jax/srt/entrypoints/openai/serving_chat.py` — disable thinking when `structural_tag` active (with debug log), fix `convert_json_schema_to_str` for constraint passing, add JSON fallback in `_process_tool_calls`
- `python/sgl_jax/srt/function_call/function_call_parser.py` — guard against `get_json_schema_constraint` returning None for non-existent function names
- `python/sgl_jax/srt/function_call/utils.py` — add missing `"type": "object"` to `_get_tool_schema` for correct json_schema constraint
- `python/sgl_jax/srt/managers/scheduler_output_processor_mixin.py` — prevent `grammar.finished` overwrite after `accept_token` failure
- `python/sgl_jax/utils.py` — use `sort_keys=True` in `convert_json_schema_to_str` for deterministic grammar cache keys

**Tests and CI:**
- `test/srt/openai_server/nightly/__init__.py` — new package
- `test/srt/openai_server/nightly/test_openai_api_surface.py` — `OpenAIApiTestParams` (frozen, with `chat()` helper) + 13 test cases + `_ResultCollector` / `_write_github_summary` for per-subtest GH summary
- `test/srt/run_suite.py` — register `nightly-test-openai-api-surface-tpu-v6e-1-daily` suite
- `.github/workflows/nightly-test-daily.yml` — new job + finish-job needs/summary sync

## Test plan

- [x] Python syntax check (`py_compile`)
- [x] YAML syntax validation
- [x] Ruff lint pass
- [x] Pre-commit hooks pass (black, isort, mypy, codespell)
- [x] `nightly-test-daily-finish` needs list includes new job (5 references verified)
- [x] `qwen25` parser confirmed in `function_call_parser.py`
- [ ] e2e constraint suite (test_json_mode, test_ebnf, test_structural_tag) passes on this branch — verifies `is_terminated` behavior change
- [ ] First nightly run on `arc-runner-v6e-1` passes all 13 tests
- [ ] `$GITHUB_STEP_SUMMARY` shows per-subtest pass/fail table

### CODE REVIEW

File: test/srt/openai_server/nightly/test_openai_api_surface.py [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+        client = self.params.make_client()
+        try:
+            response = client.chat.completions.create(
+                model=self.model,
+                messages=[{"role": "user", "content": "What is 2 + 2?"}],
+                temperature=0.0,
+                extra_body={"chat_template_kwargs": {"enable_thinking": True}},
+            )
+        except openai.APIStatusError as e:
+            self.skipTest(f"Model does not support thinking mode: {e}")
```
[the remark this answers is not in the document]
Comment (Rodrian7) [reply test/srt/openai_server/nightly/test_openai_api_surface.py:389]: Fixed in ec037ee — now re-raises on 5xx, only skips on 4xx.

File: python/sgl_jax/srt/managers/scheduler_output_processor_mixin.py [hunk 2]
```diff
@@ -241,7 +241,8 @@ def process_batch_result_prefill(
                             )
 
                             self.abort_request(AbortReq(rid=req.rid))
-                        req.grammar.finished = req.finished()
+                        else:
+                            req.grammar.finished = req.finished()
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/managers/scheduler_output_processor_mixin.py]: The effect is the same before and after modification.

Comment (Rodrian7) [reply python/sgl_jax/srt/managers/scheduler_output_processor_mixin.py]: removed the else in 388ac78f to  match sglang main.

**Review by JamesBrianD — APPROVED** [review]
